File: src/BrokerageExportProviders/Brokerages/IBKR/Transforms/Transform.py
from dataclasses import dataclass
from itertools import groupby
from typing import Generic, Sequence, TypeVar

import BrokerageExportProviders.Brokerages.IBKR.Schemas.Schemas as s
import BrokerageExportProviders.Brokerages.IBKR.Schemas.SegmentedTrades as st
import Core.FinancialEvents.Schemas.CommonFormats as cf
from Core.StagingFinancialEvents.Schemas.Events import (
    StagingTradeEvent,
    StagingTradeEventCashTransactionDividend,
    StagingTradeEventCashTransactionPaymentInLieuOfDividends,
    StagingTradeEventCashTransactionWitholdingTax,
    StagingTradeEventCashTransactionWitholdingTaxForPaymentInLieuOfDividends,
    StagingTradeEventDerivative,
    StagingTradeEventDerivativeAcquired,
    StagingTradeEventDerivativeSold,
    StagingTradeEventStock,
    StagingTradeEventStockAcquired,
    StagingTradeEventStockSold,
    StagingTransactionCash,
)
from Core.StagingFinancialEvents.Schemas.Grouping import StagingFinancialGrouping
from Core.StagingFinancialEvents.Schemas.Lots import (
    StagingTaxLot,
    StagingTaxLotMatchingDetails,
)

LINE_GENERIC_BUY = TypeVar("LINE_GENERIC_BUY")
LINE_GENERIC_SELL = TypeVar("LINE_GENERIC_SELL")


@dataclass
class SegmentedBuyEvent(Generic[LINE_GENERIC_BUY]):
    Quantity: float
    BuyLine: LINE_GENERIC_BUY


@dataclass
class SegmentedSellEvent(Generic[LINE_GENERIC_SELL]):
    Quantity: float
    SellLine: LINE_GENERIC_SELL


@dataclass
class SegmentedTradeEvents(Generic[LINE_GENERIC_BUY, LINE_GENERIC_SELL]):
    Buys: list[SegmentedBuyEvent[LINE_GENERIC_BUY]]
    Sells: list[SegmentedSellEvent[LINE_GENERIC_SELL]]


@dataclass
class SegmentedTrades(Generic[LINE_GENERIC_BUY, LINE_GENERIC_SELL]):
    Buys: list[LINE_GENERIC_BUY]
    Sells: list[LINE_GENERIC_SELL]


def convertToCashTransactions(
    cashTransactions: list[s.TransactionCash],
) -> list[StagingTransactionCash]:
    def mapToGenericDividendLine(
        transaction: s.TransactionCash,
    ) -> StagingTransactionCash:
        dividendType = cf.GenericDividendType.UNKNOWN

        ordinaryDividend = transaction.Description.lower().__contains__("ordinary dividend")
        bonusDividend = transaction.Description.lower().__contains__("bonus dividend")
        witholdingTaxForPaymentInLieuOfDividends = transaction.Description.lower().__contains__("payment in lieu of dividend")

        if ordinaryDividend:
            dividendType = cf.GenericDividendType.ORDINARY

        if bonusDividend:
            dividendType = cf.GenericDividendType.BONUS

        if transaction.Type == s.CashTransactionType.DIVIDEND:
            return StagingTradeEventCashTransactionDividend(
                ID=transaction.TransactionID,
                ISIN=transaction.ISIN,
                Ticker=transaction.Symbol,
                AssetClass=cf.GenericAssetClass.CASH_AND_CASH_EQUIVALENTS,
                Date=transaction.DateTime,
                Multiplier=1,
                DividendType=dividendType,
                ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=1,
                    UnderlyingTradePrice=transaction.Amount * transaction.FXRateToBase,  # TODO: Currency provider
                    UnderlyingCurrency=transaction.Currency,
                    ComissionCurrency=transaction.Currency,
                    ComissionTotal=0,
                    TaxCurrency=transaction.Currency,
                    TaxTotal=0,
                    FxRateToBase=transaction.FXRateToBase,
                ),
                ActionID=transaction.ActionID,
                TransactionID=transaction.TransactionID,
                ListingExchange=transaction.ListingExchange,
            )

        if transaction.Type == s.CashTransactionType.WITHOLDING_TAX and witholdingTaxForPaymentInLieuOfDividends:
            return StagingTradeEventCashTransactionWitholdingTaxForPaymentInLieuOfDividends(
                ID=transaction.TransactionID,
                ISIN=transaction.ISIN,
                Ticker=transaction.Symbol,
                AssetClass=cf.GenericAssetClass.CASH_AND_CASH_EQUIVALENTS,
                Date=transaction.DateTime,
                Multiplier=1,
                ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=1,
                    UnderlyingTradePrice=transaction.Amount * transaction.FXRateToBase,  # TODO: Currency provider
                    UnderlyingCurrency=transaction.Currency,
                    ComissionCurrency=transaction.Currency,
                    ComissionTotal=0,
                    TaxCurrency=transaction.Currency,
                    TaxTotal=0,
                    FxRateToBase=transaction.FXRateToBase,
                ),
                ActionID=transaction.ActionID,
                TransactionID=transaction.TransactionID,
                ListingExchange=transaction.ListingExchange,
            )

        if transaction.Type == s.CashTransactionType.WITHOLDING_TAX:
            return StagingTradeEventCashTransactionWitholdingTax(
                ID=transaction.TransactionID,
                ISIN=transaction.ISIN,
                Ticker=transaction.Symbol,
                AssetClass=cf.GenericAssetClass.CASH_AND_CASH_EQUIVALENTS,
                Date=transaction.DateTime,
                Multiplier=1,
                ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=1,
                    UnderlyingTradePrice=transaction.Amount * transaction.FXRateToBase,  # TODO: Currency provider
                    UnderlyingCurrency=transaction.Currency,
                    ComissionCurrency=transaction.Currency,
                    ComissionTotal=0,
                    TaxCurrency=transaction.Currency,
                    TaxTotal=0,
                    FxRateToBase=transaction.FXRateToBase,
                ),
                ActionID=transaction.ActionID,
                TransactionID=transaction.TransactionID,
                ListingExchange=transaction.ListingExchange,
            )

        if transaction.Type == s.CashTransactionType.PAYMENT_IN_LIEU_OF_DIVIDENDS:
            return StagingTradeEventCashTransactionPaymentInLieuOfDividends(
                ID=transaction.TransactionID,
                ISIN=transaction.ISIN,
                Ticker=transaction.Symbol,
                AssetClass=cf.GenericAssetClass.CASH_AND_CASH_EQUIVALENTS,
                Date=transaction.DateTime,
                Multiplier=1,
                DividendType=dividendType,
                ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=1,
                    UnderlyingTradePrice=transaction.Amount * transaction.FXRateToBase,  # TODO: Currency provider
                    UnderlyingCurrency=transaction.Currency,
                    ComissionCurrency=transaction.Currency,
                    ComissionTotal=0,
                    TaxCurrency=transaction.Currency,
                    TaxTotal=0,
                    FxRateToBase=transaction.FXRateToBase,
                ),
                ActionID=transaction.ActionID,
                TransactionID=transaction.TransactionID,
                ListingExchange=transaction.ListingExchange,
            )

        raise ValueError("Unknow type of Cash Transaction")

    return list(map(mapToGenericDividendLine, cashTransactions))


def convertStockTradesToStockTradeEvents(
    trades: Sequence[s.TradeStock],
) -> Sequence[StagingTradeEventStock]:

    def convertAcquiredTradeToAcquiredEvent(
        trade: s.TradeStock,
    ) -> StagingTradeEventStockAcquired:
        converted = StagingTradeEventStockAcquired(
            ID=trade.TransactionID,
            ISIN=trade.ISIN,
            Ticker=trade.Symbol,
            AssetClass=cf.GenericAssetClass.STOCK,
            Date=trade.DateTime,
            AcquiredReason=cf.GenericTradeReportItemGainType.BOUGHT,  # TODO: Determine reason for acquire
            Multiplier=1,
            ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                UnderlyingCurrency=trade.Currency,
                UnderlyingQuantity=trade.Quantity,
                UnderlyingTradePrice=trade.TradePrice * trade.FXRateToBase,  # TODO: Remove in favor of currency conversion provider
                ComissionCurrency=trade.IBCommissionCurrency,
                ComissionTotal=trade.IBCommission,
                TaxCurrency=trade.Currency,  # NOTE: Taxes Currency == Trade Currency ??
                TaxTotal=trade.Taxes,
                FxRateToBase=trade.FXRateToBase,
            ),
        )
        return converted

    def convertSoldTradeToSoldEvent(
        trade: s.TradeStock,
    ) -> StagingTradeEventStockSold:
        converted = StagingTradeEventStockSold(
            ID=trade.TransactionID,
            ISIN=trade.ISIN,
            Ticker=trade.Symbol,
            AssetClass=cf.GenericAssetClass.STOCK,
            Date=trade.DateTime,
            Multiplier=1,
            ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                UnderlyingCurrency=trade.Currency,
                UnderlyingQuantity=trade.Quantity,
                UnderlyingTradePrice=trade.TradePrice * trade.FXRateToBase,  # TODO: Remove in favor of currency conversion provider
                ComissionCurrency=trade.IBCommissionCurrency,
                ComissionTotal=trade.IBCommission,
                TaxCurrency=trade.Currency,  # NOTE: Taxes Currency == Trade Currency ??
                TaxTotal=trade.Taxes,
                FxRateToBase=trade.FXRateToBase,
            ),
        )
        return converted

    def convertTradeToTradeEvent(
        trade: s.TradeStock,
    ) -> StagingTradeEventStock:
        buyEvent = trade.Quantity > 0
        if buyEvent:
            return convertAcquiredTradeToAcquiredEvent(trade)
        else:
            return convertSoldTradeToSoldEvent(trade)

    tradeEvents = list(map(convertTradeToTradeEvent, trades))
    return tradeEvents


def convertStockLotsToStockLotEvents(
    lots: Sequence[s.LotStock],
) -> Sequence[StagingTaxLot]:
    def convertSingleLot(lot: s.LotStock) -> StagingTaxLot:
        converted = StagingTaxLot(
            lot.TransactionID,
            ISIN=lot.ISIN,
            Ticker=lot.Symbol,
            Quantity=lot.Quantity,
            Acquired=StagingTaxLotMatchingDetails(ID=lot.TransactionID, DateTime=None),
            Sold=StagingTaxLotMatchingDetails(ID=None, DateTime=lot.DateTime),
            ShortLongType=cf.GenericShortLong.LONG,  # TODO: Determine long / short if possible
        )
        return converted

    lotEvents = list(map(convertSingleLot, lots))
    return lotEvents


def convertDerivativeTradesToDerivativeTradeEvents(
    trades: Sequence[s.TradeDerivative],
) -> Sequence[StagingTradeEventDerivative]:

    def convertAcquiredTradeToAcquiredEvent(
        trade: s.TradeDerivative,
    ) -> StagingTradeEventDerivativeAcquired:
        converted = StagingTradeEventDerivativeAcquired(
            ID=trade.TransactionID,
            ISIN=trade.UnderlyingSecurityID,
            Ticker=trade.UnderlyingSymbol,
            AssetClass=cf.GenericAssetClass.OPTION,  # TODO: Could also be stock but leveraged (multiplier)
            Multiplier=trade.Multiplier,
            AcquiredReason=cf.GenericDerivativeReportItemGainType.BOUGHT,  # TODO: Determine reason for acquire
            Date=trade.DateTime,
            ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                UnderlyingCurrency=trade.Currency,
                UnderlyingQuantity=trade.Quantity,
                UnderlyingTradePrice=trade.TradePrice * trade.FXRateToBase,  # TODO: Remove in favor of currency conversion provider
                ComissionCurrency=trade.IBCommissionCurrency,
                ComissionTotal=trade.IBCommission,
                TaxCurrency=trade.Currency,  # NOTE: Taxes Currency == Trade Currency ??
                TaxTotal=trade.Taxes,
                FxRateToBase=trade.FXRateToBase,
            ),
        )
        return converted

    def convertSoldTradeToSoldEvent(
        trade: s.TradeDerivative,
    ) -> StagingTradeEventDerivativeSold:
        converted = StagingTradeEventDerivativeSold(
            ID=trade.TransactionID,
            ISIN=trade.UnderlyingSecurityID,
            Ticker=trade.UnderlyingSymbol,
            AssetClass=cf.GenericAssetClass.OPTION,
            Multiplier=trade.Multiplier,
            Date=trade.DateTime,
            ExchangedMoney=cf.GenericMonetaryExchangeInformation(
                UnderlyingCurrency=trade.Currency,
                UnderlyingQuantity=trade.Quantity,
                UnderlyingTradePrice=trade.TradePrice * trade.FXRateToBase,  # TODO: Remove in favor of currency conversion provider
                ComissionCurrency=trade.IBCommissionCurrency,
                ComissionTotal=trade.IBCommission,
                TaxCurrency=trade.Currency,  # NOTE: Taxes Currency == Trade Currency ??
                TaxTotal=trade.Taxes,
                FxRateToBase=trade.FXRateToBase,
            ),
        )
        return converted

    def convertTradeToTradeEvent(
        trade: s.TradeDerivative,
    ) -> StagingTradeEventDerivative:
        buyEvent = trade.Quantity > 0
        if buyEvent:
            return convertAcquiredTradeToAcquiredEvent(trade)
        else:
            return convertSoldTradeToSoldEvent(trade)

    tradeEvents = list(map(convertTradeToTradeEvent, trades))
    return tradeEvents


def convertDerivativeLotsToDerivativeLotEvents(
    lots: Sequence[s.LotDerivative],
) -> Sequence[StagingTaxLot]:
    def convertSingleLot(lot: s.LotDerivative) -> StagingTaxLot:
        converted = StagingTaxLot(
            lot.TransactionID,
            ISIN=lot.UnderlyingSecurityID,
            Ticker=lot.UnderlyingSymbol,
            Quantity=lot.Quantity,
            Acquired=StagingTaxLotMatchingDetails(ID=lot.TransactionID, DateTime=None),
            Sold=StagingTaxLotMatchingDetails(ID=None, DateTime=lot.DateTime),
            ShortLongType=cf.GenericShortLong.LONG,  # TODO: Determine long / short if possible
        )
        return converted

    lotEvents = list(map(convertSingleLot, lots))
    return lotEvents


def convertSegmentedTradesToGenericUnderlyingGroups(
    segmented: st.SegmentedTrades,
) -> Sequence[StagingFinancialGrouping]:
    stockTrades = segmented.stockTrades
    stockLots = segmented.stockLots

    derivativeTrades = segmented.derivativeTrades
    derivativeLots = segmented.derivativeLots

    cashTransactions = segmented.cashTransactions

    stockTrades.sort(key=lambda entry: entry.ISIN)
    stockLots.sort(key=lambda entry: entry.ISIN)
    derivativeTrades.sort(key=lambda entry: entry.UnderlyingSecurityID)
    derivativeLots.sort(key=lambda entry: entry.UnderlyingSecurityID)
    cashTransactions.sort(key=lambda entry: entry.ISIN)

    stockTradeEvents = convertStockTradesToStockTradeEvents(stockTrades)
    stockLotEvents = convertStockLotsToStockLotEvents(stockLots)
    cashTransactionEvents = convertToCashTransactions(cashTransactions)

    derivativeTradeEvents = convertDerivativeTradesToDerivativeTradeEvents(derivativeTrades)
    derivativeLotEvents = convertDerivativeLotsToDerivativeLotEvents(derivativeLots)

    def segmentTradeByIsin(
        trades: list[StagingTradeEvent],
    ) -> dict[str, Sequence[StagingTradeEvent]]:
        segmented: dict[str, Sequence[StagingTradeEvent]] = {}
        for key, valuesiter in groupby(trades, key=lambda trade: trade.ISIN):
            segmented[key] = list(v for v in valuesiter)
        return segmented

    def segmentLotByIsin(
        lots: list[StagingTaxLot],
    ) -> dict[str, Sequence[StagingTaxLot]]:
        segmented: dict[str, Sequence[StagingTaxLot]] = {}
        for key, valuesiter in groupby(lots, key=lambda trade: trade.ISIN):
            segmented[key] = list(v for v in valuesiter)
        return segmented

    stocksSegmented = segmentTradeByIsin(stockTradeEvents)  # type: ignore
    stockLotsSegmented = segmentLotByIsin(stockLotEvents)  # type: ignore
    derivativesSegmented = segmentTradeByIsin(derivativeTradeEvents)  # type: ignore
    derivativeLotsSegmented = segmentLotByIsin(derivativeLotEvents)  # type: ignore
    dividendsSegmented = segmentTradeByIsin(cashTransactionEvents)  # type: ignore

    allIsinsPresent = list(
        set(
            list(stocksSegmented.keys())
            + list(derivativesSegmented.keys())
            + list(stockLotsSegmented.keys())
            + list(derivativeLotsSegmented.keys())
            + list(dividendsSegmented.keys())
        )
    )

    generatedUnderlyingGroups: Sequence[StagingFinancialGrouping] = list()
    for isin in allIsinsPresent:
        wrapper = StagingFinancialGrouping(
            ISIN=isin,
            CountryOfOrigin=None,
            UnderlyingCategory=cf.GenericCategory.REGULAR,
            StockTrades=stocksSegmented.get(isin, []),  # type: ignore
            StockTaxLots=stockLotsSegmented.get(isin, []),
            DerivativeTrades=derivativesSegmented.get(isin, []),  # type: ignore
            DerivativeTaxLots=derivativeLotsSegmented.get(isin, []),
            CashTransactions=dividendsSegmented.get(isin, []),  # type: ignore
        )
        generatedUnderlyingGroups.append(wrapper)

    return generatedUnderlyingGroups

File: lib/src/Brokerages/Ibkr/Transform.ts
import {
	CashTransactionType,
	type CorporateAction,
	type LotDerivative,
	type LotStock,
	type TradeDerivative,
	type TradeStock,
	type TransactionCash,
} from "@brrr/Brokerages/Ibkr/Schemas/IbkrSchemas";
import type { SegmentedTrades } from "@brrr/Brokerages/Ibkr/Schemas/SegmentedTrades";
import {
	GenericAssetClass,
	GenericCategory,
	GenericDerivativeReportItemGainType,
	GenericDividendType,
	GenericMonetaryExchangeInformation,
	GenericShortLong,
	GenericTradeReportItemGainType,
} from "@brrr/Core/Schemas/CommonFormats";
import {
	StagingTradeEventCashTransactionDividend,
	StagingTradeEventCashTransactionPaymentInLieuOfDividends,
	StagingTradeEventCashTransactionWithholdingTax,
	StagingTradeEventCashTransactionWithholdingTaxForPaymentInLieuOfDividends,
	StagingTradeEventDerivativeAcquired,
	StagingTradeEventDerivativeSold,
	StagingTradeEventStockAcquired,
	StagingTradeEventStockSold,
} from "@brrr/Core/Schemas/Staging/Events";
import { StagingFinancialGrouping } from "@brrr/Core/Schemas/Staging/Grouping";
import {
	StagingIdentifierChangeType,
	StagingIdentifierRelationshipPartial,
	type StagingIdentifierRelationshipPartialAny,
	StagingIdentifierRelationshipPartialWithQuantity,
	StagingIdentifierRelationships,
} from "@brrr/Core/Schemas/Staging/IdentifierRelationship";
import { StagingTaxLot, StagingTaxLotMatchingDetails } from "@brrr/Core/Schemas/Staging/Lots";
import { StagingFinancialEvents } from "@brrr/Core/Schemas/Staging/StagingFinancialEvents";
import { StagingFinancialIdentifier } from "@brrr/Core/Schemas/Staging/StagingFinancialIdentifier";

type AnyEvent = { financialIdentifier: StagingFinancialIdentifier };

export class IbkrTransformService {
	public convertSegmentedTradesToStagingEvents(segmented: SegmentedTrades): StagingFinancialEvents {
		const stockTrades = [...segmented.stockTrades].sort((a, b) => a.isin.localeCompare(b.isin));
		const stockLots = [...segmented.stockLots].sort((a, b) => a.isin.localeCompare(b.isin));
		const derivativeTrades = [...segmented.derivativeTrades].sort((a, b) =>
			a.underlyingSecurityID.localeCompare(b.underlyingSecurityID)
		);
		const derivativeLots = [...segmented.derivativeLots].sort((a, b) => a.underlyingSecurityID.localeCompare(b.underlyingSecurityID));
		const cashTransactions = [...segmented.cashTransactions].sort((a, b) => a.isin.localeCompare(b.isin));

		const stockTradeEvents = this._convertStockTradesToEvents(stockTrades);
		const stockLotEvents = this._convertStockLotsToEvents(stockLots);
		const cashTransactionEvents = this._convertToCashTransactions(cashTransactions);
		const derivativeTradeEvents = this._convertDerivativeTradesToEvents(derivativeTrades);
		const derivativeLotEvents = this._convertDerivativeLotsToEvents(derivativeLots);

		const groupByIdentifier = <T extends AnyEvent>(items: T[]): Map<string, { identifier: StagingFinancialIdentifier; items: T[] }> => {
			const map = new Map<string, { identifier: StagingFinancialIdentifier; items: T[] }>();
			for (const item of items) {
				const key = item.financialIdentifier.toKey();
				if (!map.has(key)) {
					map.set(key, { identifier: item.financialIdentifier, items: [] });
				}
				map.get(key)!.items.push(item);
			}
			return map;
		};

		const stocksMap = groupByIdentifier(stockTradeEvents);
		const stockLotsMap = groupByIdentifier(stockLotEvents);
		const derivativesMap = groupByIdentifier(derivativeTradeEvents);
		const derivativeLotsMap = groupByIdentifier(derivativeLotEvents);
		const dividendsMap = groupByIdentifier(cashTransactionEvents);

		const allKeys = new Set([
			...stocksMap.keys(),
			...stockLotsMap.keys(),
			...derivativesMap.keys(),
			...derivativeLotsMap.keys(),
			...dividendsMap.keys(),
		]);

		const groupings: StagingFinancialGrouping[] = [];
		for (const key of allKeys) {
			const identifier = (
				stocksMap.get(key) ??
					stockLotsMap.get(key) ??
					derivativesMap.get(key) ??
					derivativeLotsMap.get(key) ??
					dividendsMap.get(key)
			)!.identifier;

			groupings.push(
				new StagingFinancialGrouping({
					financialIdentifier: identifier,
					countryOfOrigin: null,
					underlyingCategory: GenericCategory.REGULAR,
					stockTrades: (stocksMap.get(key)?.items ?? []) as (StagingTradeEventStockAcquired | StagingTradeEventStockSold)[],
					stockTaxLots: stockLotsMap.get(key)?.items ?? [],
					derivativeTrades:
						(derivativesMap.get(key)?.items ?? []) as (StagingTradeEventDerivativeAcquired | StagingTradeEventDerivativeSold)[],
					derivativeTaxLots: derivativeLotsMap.get(key)?.items ?? [],
					cashTransactions: dividendsMap.get(key)?.items ?? [],
				}),
			);
		}

		const partials = this._convertCorporateActionsToPartialRelationships(segmented.corporateActions);
		const identifierRelationships = new StagingIdentifierRelationships({
			relationships: [],
			partialRelationships: partials,
		});

		return new StagingFinancialEvents({ groupings, identifierRelationships });
	}

	private _convertToCashTransactions(cashTransactions: TransactionCash[]) {
		return cashTransactions.map((transaction) => {
			const descLower = transaction.description.toLowerCase();
			const isOrdinaryDividend = descLower.includes("ordinary dividend");
			const isBonusDividend = descLower.includes("bonus dividend");
			const isPaymentInLieuWithholding = descLower.includes("payment in lieu of dividend");

			let dividendType = GenericDividendType.UNKNOWN;
			if (isOrdinaryDividend) dividendType = GenericDividendType.ORDINARY;
			if (isBonusDividend) dividendType = GenericDividendType.BONUS;

			const financialIdentifier = new StagingFinancialIdentifier({ isin: transaction.isin, ticker: transaction.symbol });
			const exchangedMoney = new GenericMonetaryExchangeInformation({
				underlyingQuantity: 1,
				underlyingTradePrice: transaction.amount * transaction.fxRateToBase,
				underlyingCurrency: transaction.currency,
				comissionCurrency: transaction.currency,
				comissionTotal: 0,
				taxCurrency: transaction.currency,
				taxTotal: 0,
				fxRateToBase: transaction.fxRateToBase,
			});
			const base = {
				id: transaction.transactionID,
				financialIdentifier,
				assetClass: GenericAssetClass.CASH_AND_CASH_EQUIVALENTS,
				date: transaction.dateTime,
				multiplier: 1,
				exchangedMoney,
				actionId: transaction.actionID,
				transactionId: transaction.transactionID,
				listingExchange: transaction.listingExchange,
			};

			if (transaction.type === CashTransactionType.DIVIDEND) {
				return new StagingTradeEventCashTransactionDividend({ ...base, dividendType });
			}

			if (transaction.type === CashTransactionType.WITHHOLDING_TAX && isPaymentInLieuWithholding) {
				return new StagingTradeEventCashTransactionWithholdingTaxForPaymentInLieuOfDividends(base);
			}

			if (transaction.type === CashTransactionType.WITHHOLDING_TAX) {
				return new StagingTradeEventCashTransactionWithholdingTax(base);
			}

			if (transaction.type === CashTransactionType.PAYMENT_IN_LIEU_OF_DIVIDENDS) {
				return new StagingTradeEventCashTransactionPaymentInLieuOfDividends({ ...base, dividendType });
			}

			throw new Error(`Unknown cash transaction type: ${transaction.type}`);
		});
	}

	private _convertStockTradesToEvents(trades: TradeStock[]) {
		return trades.map((trade) => {
			const financialIdentifier = new StagingFinancialIdentifier({ isin: trade.isin, ticker: trade.symbol });
			const exchangedMoney = new GenericMonetaryExchangeInformation({
				underlyingCurrency: trade.currency,
				underlyingQuantity: trade.quantity,
				underlyingTradePrice: trade.tradePrice * trade.fxRateToBase,
				comissionCurrency: trade.ibCommissionCurrency,
				comissionTotal: trade.ibCommission * trade.fxRateToBase,
				taxCurrency: trade.currency, // NOTE: Taxes Currency == Trade Currency ??
				taxTotal: trade.taxes,
				fxRateToBase: trade.fxRateToBase,
			});
			const base = {
				id: trade.transactionID,
				financialIdentifier,
				assetClass: GenericAssetClass.STOCK,
				date: trade.dateTime,
				multiplier: 1,
				exchangedMoney,
			};
			if (trade.quantity > 0) {
				return new StagingTradeEventStockAcquired({ ...base, acquiredReason: GenericTradeReportItemGainType.BOUGHT });
			} else {
				return new StagingTradeEventStockSold(base);
			}
		});
	}

	private _convertStockLotsToEvents(lots: LotStock[]): StagingTaxLot[] {
		return lots.map((lot) =>
			new StagingTaxLot({
				id: lot.transactionID,
				financialIdentifier: new StagingFinancialIdentifier({ isin: lot.isin, ticker: lot.symbol }),
				quantity: lot.quantity,
				acquired: new StagingTaxLotMatchingDetails({ id: lot.transactionID, dateTime: null }),
				sold: new StagingTaxLotMatchingDetails({ id: null, dateTime: lot.dateTime }),
				shortLongType: GenericShortLong.LONG,
			})
		);
	}

	private _convertDerivativeTradesToEvents(trades: TradeDerivative[]) {
		return trades.map((trade) => {
			const financialIdentifier = new StagingFinancialIdentifier({
				isin: trade.underlyingSecurityID,
				ticker: trade.symbol,
				name: trade.description,
			});
			const exchangedMoney = new GenericMonetaryExchangeInformation({
				underlyingCurrency: trade.currency,
				underlyingQuantity: trade.quantity,
				underlyingTradePrice: trade.tradePrice * trade.fxRateToBase,
				comissionCurrency: trade.ibCommissionCurrency,
				comissionTotal: trade.ibCommission * trade.fxRateToBase,
				taxCurrency: trade.currency, // NOTE: Taxes Currency == Trade Currency ??
				taxTotal: trade.taxes,
				fxRateToBase: trade.fxRateToBase,
			});
			const base = {
				id: trade.transactionID,
				financialIdentifier,
				assetClass: GenericAssetClass.OPTION,
				multiplier: trade.multiplier,
				date: trade.dateTime,
				exchangedMoney,
			};
			if (trade.quantity > 0) {
				return new StagingTradeEventDerivativeAcquired({ ...base, acquiredReason: GenericDerivativeReportItemGainType.BOUGHT });
			} else {
				return new StagingTradeEventDerivativeSold(base);
			}
		});
	}

	private _convertDerivativeLotsToEvents(lots: LotDerivative[]): StagingTaxLot[] {
		return lots.map((lot) =>
			new StagingTaxLot({
				id: lot.transactionID,
				financialIdentifier: new StagingFinancialIdentifier({
					isin: lot.underlyingSecurityID,
					ticker: lot.symbol,
					name: lot.description,
				}),
				quantity: lot.quantity,
				acquired: new StagingTaxLotMatchingDetails({ id: lot.transactionID, dateTime: null }),
				sold: new StagingTaxLotMatchingDetails({ id: null, dateTime: lot.dateTime }),
				shortLongType: GenericShortLong.LONG,
			})
		);
	}

	private _convertCorporateActionsToPartialRelationships(corporateActions: CorporateAction[]): StagingIdentifierRelationshipPartialAny[] {
		const partials: StagingIdentifierRelationshipPartialAny[] = [];

		for (const row of corporateActions) {
			const identifier = new StagingFinancialIdentifier({ isin: row.isin, ticker: row.symbol });
			const isFromSide = row.symbol.endsWith(".OLD");
			const actionType = (row.type || "").toUpperCase();

			let changeType: StagingIdentifierChangeType;
			// IBKR corporate action type codes: "IC" = ticker/identifier change (RENAME), "FI" = stock split (SPLIT)
			if (actionType === "IC") {
				changeType = StagingIdentifierChangeType.RENAME;
			} else if (actionType === "FI") {
				changeType = StagingIdentifierChangeType.SPLIT;
			} else {
				changeType = StagingIdentifierChangeType.UNKNOWN;
			}

			const fromIdentifier = isFromSide ? identifier : null;
			const toIdentifier = isFromSide ? null : identifier;

			if (changeType === StagingIdentifierChangeType.SPLIT) {
				partials.push(
					new StagingIdentifierRelationshipPartialWithQuantity({
						fromIdentifier,
						toIdentifier,
						correlationKey: row.actionID,
						changeType,
						effectiveDate: row.dateTime,
						quantity: Math.abs(row.quantity),
					}),
				);
			} else {
				partials.push(
					new StagingIdentifierRelationshipPartial({
						fromIdentifier,
						toIdentifier,
						correlationKey: row.actionID,
						changeType,
						effectiveDate: row.dateTime,
					}),
				);
			}
		}

		return partials;
	}
}
